File: python/say/say.py
# coding=utf-8
from __future__ import unicode_literals

__author__ = 'Cedric Zhuang'

ones = ('zero one two three four '
        'five six seven eight nine').split()

teens = ('ten eleven twelve thirteen fourteen '
         'fifteen sixteen seventeen eighteen nineteen').split()

tens = ('n/a n/a twenty thirty forty '
        'fifty sixty seventy eighty ninety').split()


def say(n):
    if n == 0:
        ret = ones[n]
    else:
        ret = _say(n)
    return ret.strip()


def _say(n, prefix=''):
    if n < 100 and prefix:
        ret = prefix
    else:
        ret = ''

    n = int(n)
    if n < 0:
        raise AttributeError('n must be a positive number.')
    elif n == 0:
        ret = ''
    elif n < 10:
        ret += ones[n]
    elif n < 20:
        ret += teens[n - 10]
    elif n < 100:
        ret += tens[n / 10] + _say(n % 10, '-')
    elif n < 1000:
        ret += _unit(n, 100, 'hundred')
    elif n < 1e6:
        ret = _unit(n, 1e3, 'thousand')
    elif n < 1e9:
        ret = _unit(n, 1e6, 'million')
    elif n < 1e12:
        ret = _unit(n, 1e9, 'billion')
    else:
        raise AttributeError('n must be less than or equal to 1e12.')
    return ret


def _unit(n, l, word):
    x, y = divmod(n, l)
    return '%s %s %s' % (_say(x), word, _say(y, 'and '))

File: cpp/say/say.cpp
#include "say.h"
#include <cmath>
#include <string>
#include <sstream>
#include <map>
#include <exception>
#include <boost/algorithm/string.hpp>


namespace say
{

using std::string;

static const string ones[] = { 
    "", "one", "two", "three", "four",
    "five", "six", "seven", "eight", "nine" 
};

static const string teens[] = {
    "ten", "eleven", "twelve", "thirteen", "fourteen",
    "fifteen", "sixteen", "seventeen", "eighteen", "nineteen"
};

static const string tens[] = {
    "", "", "twenty", "thirty", "forty",
    "fifty", "sixty", "seventy", "eighty", "ninty"
};

static const string huge[] = {
    "", "thousand", "million", "billion"
};

string _nz_in_english(const ull& n)
{
    string ret;
    if (n < 10)
    {
        ret = ones[n];
    }
    else if (n < 20)
    {
        ret = teens[n - 10];
    }
    else if (n < 100)
    {
        ret = tens[n / 10];
        ull single = n % 10;
        if (single)
        {
            ret += '-' + _nz_in_english(single);
        }
    }
    else if (n < 1000)
    {
        ret = ones[n / 100] + " hundred " + _nz_in_english(n % 100);
    }
    return ret;
}

string in_english(ull n)
{
    std::ostringstream ret;
    if (n < 0 || n >= pow(1000, 4))
    {
        throw std::domain_error("n should be larger than 0.");
    }
    else if (n == 0)
    {
        ret << "zero";
    }
    else
    {
        for (int i = 3; i >= 0; i--)
        {
            ull base = (ull)pow(1000, i);
            if (n / base)
            {
                ret << _nz_in_english(n / base) << ' ' << huge[i] << ' ';
            }
            n %= base;
        }
    }
    return boost::trim_copy(ret.str());
}


} // namespace say